Incorrectly formats XML files with DOCTYPE statements

See RENCI-NRIG/orca5#176

[Example diff](https://github.com/RENCI-NRIG/orca5/commit/0ff070ee9df1293f74b9f16e26f2b33f0f38a71e#diff-6d0ca63a6a8548b23def2bbe58b2cbfd)
``` 
-<!DOCTYPE project [
-<!ENTITY deps SYSTEM “ant/deps.xml”>
-<!ENTITY core SYSTEM “ant/core.xml”>
-<!ENTITY packages SYSTEM “ant/packages.xml”>
-<!ENTITY tests SYSTEM “ant/tests.xml”>
-]>

+<!DOCTYPE project> 
+<!--ENTITY core SYSTEM “ant/core.xml”--> 
+<!--ENTITY packages SYSTEM “ant/packages.xml”--> 
+<!--ENTITY tests SYSTEM “ant/tests.xml”-->
+ ]&gt;
```

I'm not involved with that project anymore, but I took a quick look anyway.

First, it's not immediately obvious how I start using this new xml-formatter project.  Do I just need to use the latest version of the `formatter-maven-plugin` plugin?

Unfortunately that resulted in some SAXParserException on this (orca5) project:
> [ERROR] Failed to execute goal net.revelc.code.formatter:formatter-maven-plugin:2.9.0:format (default-cli) on project orca: Execution default-cli of goal net.revelc.code.formatter:formatter-maven-plugin:2.9.0:format failed: org.xml.sax.SAXParseException; lineNumber: 9; columnNumber: 2; The markup in the document following the root element must be well-formed. -> [Help 1]


The error message doesn't tell you which file caused the error, but turning on maven debugging (`-X`) seems to indicate that it was processing [orca5/blob/master/boot/ant/deps.xml](https://github.com/RENCI-NRIG/orca5/blob/master/boot/ant/deps.xml), although I don't see any obvious errors in that file.

--- a/src/main/java/net/revelc/code/formatter/xml/lib/FormattingPreferences.java
+++ b/src/main/java/net/revelc/code/formatter/xml/lib/FormattingPreferences.java
@@ -90,10 +90,10 @@ public class FormattingPreferences {
     public String getSaxValidation() { return saxValidation; }
 
     public void setSaxValidation(String saxValidation) {
-        if (saxValidation.equals(IGNORE) || saxValidation.equals(FAIL)) {
+        if (saxValidation.equals(IGNORE) || saxValidation.equals(FAIL) || saxValidation.equals(WARN)) {
             this.saxValidation = saxValidation;
         } else {
-            this.saxValidation = WARN;
+            throw new IllegalArgumentException("Invalid configuration value for SAX Validation: " + saxValidation);
         }
     }
 }

--- a/src/main/java/net/revelc/code/formatter/xml/lib/FormattingPreferences.java
+++ b/src/main/java/net/revelc/code/formatter/xml/lib/FormattingPreferences.java
@@ -16,11 +16,16 @@
 package net.revelc.code.formatter.xml.lib;
 
 public class FormattingPreferences {
+    public static final String IGNORE = "IGNORE";
+    public static final String WARN = "WARN";
+    public static final String FAIL = "FAIL";
+
     private int maxLineLength = 120;
     private boolean wrapLongLines = true;
     private boolean tabInsteadOfSpaces = true;
     private int tabWidth = 4;
     private boolean splitMultiAttrs = false;
+    private String saxValidation = WARN;
 
     public void setMaxLineLength(Integer maxLineLength) {
         if (maxLineLength != null)
@@ -80,5 +85,15 @@ public class FormattingPreferences {
     public void setSplitMultiAttrs(Boolean setSplitMultiAttrs) {
         if (setSplitMultiAttrs != null)
             this.splitMultiAttrs = setSplitMultiAttrs;
+    }
+
+    public String getSaxValidation() { return saxValidation; }
+
+    public void setSaxValidation(String saxValidation) {
+        if (saxValidation.equals(IGNORE) || saxValidation.equals(FAIL)) {
+            this.saxValidation = saxValidation;
+        } else {
+            this.saxValidation = WARN;
+        }
     }
 }

--- a/src/main/java/net/revelc/code/formatter/xml/lib/XmlDocumentFormatter.java
+++ b/src/main/java/net/revelc/code/formatter/xml/lib/XmlDocumentFormatter.java
@@ -117,7 +117,7 @@ public class XmlDocumentFormatter {
             reader.parse(new InputSource(new StringReader(documentText)));
         } catch (Exception exception) {
             if (prefs.getSaxValidation().equals(FormattingPreferences.WARN)) {
-                System.out.println(exception.getMessage());
+                System.err.println("WARN: " + exception.getMessage());
             } else {
                 throw new IllegalArgumentException(exception);
             }

--- a/src/main/java/net/revelc/code/formatter/xml/lib/XmlDocumentFormatter.java
+++ b/src/main/java/net/revelc/code/formatter/xml/lib/XmlDocumentFormatter.java
@@ -74,7 +74,9 @@ public class XmlDocumentFormatter {
     }
 
     public String format(String documentText) {
-        validateWellFormedness(documentText);
+        if (!prefs.getSaxValidation().equals(FormattingPreferences.IGNORE)) {
+            validateWellFormedness(documentText);
+        }
 
         Reader reader = new StringReader(documentText);
         formattedXml = new StringBuilder();
@@ -114,7 +116,11 @@ public class XmlDocumentFormatter {
             reader.setErrorHandler(errorHandler);
             reader.parse(new InputSource(new StringReader(documentText)));
         } catch (Exception exception) {
-            throw new IllegalArgumentException(exception);
+            if (prefs.getSaxValidation().equals(FormattingPreferences.WARN)) {
+                System.out.println(exception.getMessage());
+            } else {
+                throw new IllegalArgumentException(exception);
+            }
         }
     }
 

--- a/src/test/java/net/revelc/code/formatter/xml/lib/FormatterTest.java
+++ b/src/test/java/net/revelc/code/formatter/xml/lib/FormatterTest.java
@@ -79,4 +79,15 @@ public class FormatterTest {
 
         assertDoesNotThrow(() -> formatter.format(inXml));
     }
+
+    @Test
+    public void testNoRootElementFails() throws Exception {
+        FormattingPreferences prefs = new FormattingPreferences();
+        prefs.setSaxValidation(FormattingPreferences.FAIL);
+        XmlDocumentFormatter formatter = new XmlDocumentFormatter(System.lineSeparator(), prefs);
+        String inXml = new String(Files.readAllBytes(Paths.get("src/test/resources/sample-orca5-deps.xml")),
+                StandardCharsets.UTF_8);
+
+        assertThrows(IllegalArgumentException.class, () -> formatter.format(inXml));
+    }
 }

--- a/src/test/java/net/revelc/code/formatter/xml/lib/FormatterTest.java
+++ b/src/test/java/net/revelc/code/formatter/xml/lib/FormatterTest.java
@@ -11,6 +11,7 @@ package net.revelc.code.formatter.xml.lib;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
@@ -68,5 +69,14 @@ public class FormatterTest {
                 StandardCharsets.UTF_8);
 
         assertThrows(IllegalArgumentException.class, () -> formatter.format(inXml));
+    }
+
+    @Test
+    public void testNoRootElement() throws Exception {
+        XmlDocumentFormatter formatter = new XmlDocumentFormatter(System.lineSeparator(), new FormattingPreferences());
+        String inXml = new String(Files.readAllBytes(Paths.get("src/test/resources/sample-orca5-deps.xml")),
+                StandardCharsets.UTF_8);
+
+        assertDoesNotThrow(() -> formatter.format(inXml));
     }
 }

--- a/src/test/resources/sample-orca5-deps.xml
+++ b/src/test/resources/sample-orca5-deps.xml
@@ -1,0 +1,23 @@
+<!-- from https://github.com/RENCI-NRIG/orca5/blob/master/boot/ant/deps.xml -->
+<!--
+    User-specific project build properties. Properties defined in this file have
+    precedence over properties defined in build.properties
+-->
+<property file="${basedir}/ant/user.properties" />
+        <!--
+             Project buld properties.
+        -->
+<property file="${basedir}/ant/build.properties" />
+        <!--
+            User-specific project test properties. Properties defined in this file have
+            precedence over properties defined in tests.properties
+        -->
+<property file="${basedir}/ant/user.tests.properties" />
+        <!--
+             Project test properties.
+        -->
+<property file="${basedir}/ant/tests.properties" />
+        <!-- Antcontrib tasks -->
+<taskdef resource="net/sf/antcontrib/antlib.xml" classpath="${compile_classpath}" />
+        <!-- xjc task: used for generating beans from xsd files -->
+<taskdef name="xjc" classname="com.sun.tools.xjc.XJCTask" classpath="${compile_classpath}" />

